pkg/terraform/state/handler.go

```diff
@@ -76,136 +73,178 @@ func (h *handler) OnChange(key string, obj *v1.State) (*v1.State, error) {
 		return nil, nil
 	}
 
-	if obj.Spec.Version < 1 {
-		obj.Spec.Version = 1
-	}
-	if obj.Spec.Image == "" {
-		obj.Spec.Image = fmt.Sprintf("%s:latest", DefaultExecutorImage)
-	}
-
 	input, ok, err := h.gatherInput(obj)
 	if err != nil {
 		return obj, err
 	}
+
 	if !ok {
 		v1.ExecutionConditionMissingInfo.SetStatus(obj, err.Error())
 		return h.states.Update(obj)
 	}
 
-	v1.ExecutionConditionMissingInfo.False(obj)
-	v1.ExecutionConditionJobDeployed.False(obj)
+	if v1.StateConditionJobDeployed.IsTrue(obj) && obj.Status.LastRunHash != "" {
+		return obj, nil
+	}
 
-	err = v1.ExecutionConditionJobDeployed.DoUntilTrue(obj, func() (runtime.Object, error) {
-		runName, err := h.deployCreate(obj, input, ActionCreate)
-		logrus.Debugf("Execution Name: %s", runName)
-		if err != nil {
-			return obj, err
-		}
+	runHash := createRunHash(obj, input, ActionCreate)
+	if runHash == obj.Status.LastRunHash {
+		return obj, nil
+	}
 
-		if obj.Status.ExecutionName != runName {
-			obj.Status.ExecutionName = runName
-			return h.states.Update(obj)
-		}
+	//running an execution
+	obj.Status.LastRunHash = runHash
+	obj, err = h.states.Update(obj)
+	if err != nil {
+		logrus.Debug(err)
+		return obj, nil
+	}
 
-		v1.ExecutionConditionJobDeployed.True(obj)
+	logrus.Debug("lock acquired with new hash")
 
-		return obj, nil
-	})
+	if obj.Spec.Version < 1 {
+		obj.Spec.Version = 1
+	}
+	if obj.Spec.Image == "" {
+		obj.Spec.Image = fmt.Sprintf("%s:latest", DefaultExecutorImage)
+	}
 
+	//new execution if none running
+	exec, err := h.deployCreate(obj, input)
 	if err != nil {
+		logrus.Debug(err)
```

Comment:
A debug log should have some context, random errors will be hard to trace

Reply:
agreed added context.